skyportal/models.py

```diff
@@ -1645,39 +1652,170 @@ class Spectrum(Base):
+           in the input file this value will be ignored. If there are more than
+           2 columns in the input file, but none of them correspond to flux error
+           values, set this parameter to None.
 
         Returns
         -------
-
         spec : `skyportal.models.Spectrum`
            The Spectrum generated from the ASCII file.
 
         """
-        data = np.loadtxt(filename)
-        if data.shape[1] != 2:  # TODO support other formats
-            raise ValueError(f"Expected 2 columns, got {data.shape[1]}")
+
+        filename = Path(filename)
+
+        if data is None:
+            try:
+                table = ascii.read(filename, comment='#', header_start=None)
+            except Exception as e:
+                e.args = (f'Error parsing ASCII file: {e.args[0]}',)
+                raise
+            bytes = open(filename).read().encode('ascii')
+        else:
+            file_obj = io.BytesIO(data.encode('ascii'))
+            try:
+                table = ascii.read(file_obj, comment='#', header_start=None)
+            except Exception as e:
+                e.args = (f'Error parsing ASCII file: {e.args[0]}',)
+                raise
+            bytes = data
+
+        tabledata = np.asarray(table)
+        colnames = table.colnames
+
+        if len(colnames) < 2:
+            raise ValueError(
+                'Input data must have at least 2 columns (wavelength, '
+                'flux, and optionally flux error).'
+            )
+
+        if 'comments' in table.meta:
+
+            # this section matches lines like:
+            # XTENSION: IMAGE
+            # BITPIX: -32
+            # NAXIS: 2
+            # NAXIS1: 433
+            # NAXIS2: 1
+
+            header = {}
+            for line in table.meta['comments']:
+                try:
+                    result = yaml.load(line, Loader=yaml.FullLoader)
+                except yaml.YAMLError:
+                    continue
+                if isinstance(result, dict):
+                    header.update(result)
+
+            # otherwise read it like a fits header
+            cards = []
+            with warnings.catch_warnings():
+                warnings.simplefilter('error', AstropyWarning)
+                for line in table.meta['comments']:
+                    # this line does not raise a warning
+                    card = fits.Card.fromstring(line)
+                    try:
+                        # this line warns (exception in this context)
+                        card.verify()
+                    except AstropyWarning:
+                        continue
+                    cards.append(card)
+
+            # this ensures lines like COMMENT and HISTORY are properly dealt with
+            fits_header = fits.Header(cards=cards)
+            serialized = dict(fits_header)
+
+            for key in serialized:
+                if isinstance(serialized[key], fits.header._CardAccessor):
```

Comment:
This doesn't seem safe; checking for a private type.

Reply:
Fixed this to be safer.